### PR TITLE
Changes to handle REJECTED_TIMEDOUT

### PR DESCRIPTION
**Pull request by favalos**
Changes to handle #206.

### CODE REVIEW

**Comment by favalos** [comment]
@dfarr Please let me know if that is what you have in mind.

I had to do some changes to the generator and model.

Let me know if any suggestions.

Thanks!

File: internal/app/coroutines/completePromise.go [hunk 1]
```diff
@@ -165,8 +165,9 @@ func CompletePromise(metadata *metadata.Metadata, req *t_api.Request, res func(*
 			} else {
 				status := t_api.ForbiddenStatus(p.State)
 				strict := req.CompletePromise.Strict && p.State != req.CompletePromise.State
+				timeout := !req.CompletePromise.Strict && req.CompletePromise.State.In(promise.Timedout)
```
Comment (dfarr) [inline internal/app/coroutines/completePromise.go:168]:
```suggestion
				timeout := !req.CompletePromise.Strict && req.CompletePromise.State == promise.Timedout
```

Since we're only checking for one state, we can do a direct comparison

Comment (dfarr) [inline internal/app/coroutines/completePromise.go]:
```suggestion
				timeout := !req.CompletePromise.Strict && p.State == promise.Timedout
```

We need to check the state of the promise, as opposed to the state in the request

File: internal/app/subsystems/api/http/promise.go [hunk 2]
```diff
@@ -121,7 +121,7 @@ func (s *server) completePromise(c *gin.Context) {
 		err  error
 	)
 
-	if !body.State.In(promise.Resolved | promise.Rejected | promise.Canceled) {
+	if !body.State.In(promise.Resolved | promise.Rejected | promise.Canceled | promise.Timedout) {
```
Comment (dfarr) [inline internal/app/subsystems/api/http/promise.go]:
```suggestion
	if !body.State.In(promise.Resolved | promise.Rejected | promise.Canceled) {
```

A promise can only timeout when tick > promise.Timeout, you cannot time out a promise on request.

File: test/dst/dst.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ func (d *DST) Run(r *rand.Rand, api api.API, aio aio.AIO, system *system.System,
 			generator.AddRequest(generator.GenerateCreatePromise)
 			model.AddResponse(t_api.CreatePromise, model.ValidatCreatePromise)
 		case t_api.CompletePromise:
-			generator.AddRequest(generator.GenerateCancelPromise)
+			generator.AddRequest(generator.GenerateCompletePromise)
```
Comment (dfarr) [inline test/dst/dst.go:61]: Nice catch!

File: test/dst/generator.go [hunk 4]
```diff
@@ -193,62 +194,19 @@ func (g *Generator) GenerateCreatePromise(r *rand.Rand, t int64) *t_api.Request
...
 }
 
-func (g *Generator) GenerateCancelPromise(r *rand.Rand, t int64) *t_api.Request {
+func (g *Generator) GenerateCompletePromise(r *rand.Rand, t int64) *t_api.Request {
 	id := g.idSet[r.Intn(len(g.idSet))]
 	idempotencyKey := g.idemotencyKeySet[r.Intn(len(g.idemotencyKeySet))]
 	data := g.dataSet[r.Intn(len(g.dataSet))]
 	headers := g.headersSet[r.Intn(len(g.headersSet))]
 	strict := r.Intn(2) == 0
+	state := promise.State(math.Exp2(float64(r.Intn(4) + 1)))
```
Comment (dfarr) [inline test/dst/generator.go]: Oh nice this is very clever! The current state enum looks like this
```go
const (
	Pending State = 1 << iota
	Resolved // 2
	Rejected // 4
	Timedout // 8
	Canceled // 16
)
```

I recommend swapping Timedout and Canceled so that the 3 valid states for the `CompletePromise` request will be in order. That way we can still use this random approach, also it feels better to me have Timedout last for some reason - what do you think?

**Comment by dfarr** [comment]
Looks good @favalos! I left a few comments, the main thing is to not permit a request to timeout a promise, only the server can time out a promise.

File: internal/app/coroutines/completePromise.go (hunk 1 above)
Comment (favalos) [reply internal/app/coroutines/completePromise.go]: This makes sense, thanks for clarifying that timeout must not be allowed on request.

File: internal/app/subsystems/api/http/promise.go (hunk 2 above)
Comment (favalos) [reply internal/app/subsystems/api/http/promise.go]: Thanks, I did the changes accordingly.

File: test/dst/generator.go (hunk 4 above)
Comment (favalos) [reply test/dst/generator.go]: Yes, I think that order makes more sense. 

Working on it.

Comment (favalos) [reply test/dst/generator.go]: Done, thanks!

**Comment by favalos** [comment]
Thanks for the clarification @dfarr, I completed the changes accordingly.

Let me know if any other comments.

File: test/dst/model.go [hunk 5]
```diff
@@ -268,7 +269,13 @@ func (m *Model) ValidateCompletePromise(t int64, req *t_api.Request, res *t_api.
 		pm.promise = res.CompletePromise.Promise
 		return nil
 	case t_api.StatusCreated:
-		if res.CompletePromise.Promise.State != promise.Canceled {
+		if req.CompletePromise.State.In(promise.Resolved) && res.CompletePromise.Promise.State != promise.Resolved {
```
Comment (dfarr) [inline test/dst/model.go]:
```suggestion
		if req.CompletePromise.State == promise.Resolved && res.CompletePromise.Promise.State != promise.Resolved {
```

File: test/dst/model.go [hunk 6]
```diff
@@ -268,7 +269,13 @@ func (m *Model) ValidateCompletePromise(t int64, req *t_api.Request, res *t_api.
...
+			return fmt.Errorf("unexpected state %s after resolve promise", res.CompletePromise.Promise.State)
+		}
+		if req.CompletePromise.State.In(promise.Rejected) && res.CompletePromise.Promise.State != promise.Rejected {
```
Comment (dfarr) [inline test/dst/model.go]:
```suggestion
		if req.CompletePromise.State == promise.Rejected && res.CompletePromise.Promise.State != promise.Rejected {
```

File: test/dst/model.go [hunk 7]
```diff
@@ -268,7 +269,13 @@ func (m *Model) ValidateCompletePromise(t int64, req *t_api.Request, res *t_api.
...
+			return fmt.Errorf("unexpected state %s after reject promise", res.CompletePromise.Promise.State)
+		}
+		if req.CompletePromise.State.In(promise.Canceled) && res.CompletePromise.Promise.State != promise.Canceled {
```
Comment (dfarr) [inline test/dst/model.go]:
```suggestion
		if req.CompletePromise.State == promise.Canceled && res.CompletePromise.Promise.State != promise.Canceled {
```

File: test/dst/model.go [hunk 8]
```diff
@@ -252,9 +252,10 @@ func (m *Model) ValidateCompletePromise(t int64, req *t_api.Request, res *t_api.
 	switch res.CompletePromise.Status {
 	case t_api.StatusOK:
 		if pm.completed() {
-			if !pm.idempotencyKeyForCompleteMatch(res.CompletePromise.Promise) {
+			if !(!req.CompletePromise.Strict && pm.promise.State == promise.Timedout) &&
```
Comment (dfarr) [inline test/dst/model.go]: I'm finding this if statement a little hard to follow, so I plugged it into a program that simplifies boolean expressions. To me the result is a little easier to understand, let me know what you think.

```
!(!req.CompletePromise.Strict && pm.promise.State == promise.Timedout)
  && !pm.idempotencyKeyForCompleteMatch(res.CompletePromise.Promise)

a = req.CompletePromise.Strict
b = pm.promise.State == promise.Timedout
c = pm.idempotencyKeyForCompleteMatch(res.CompletePromise.Promise)

!(!a && b) && !c
!c && !(!a && b)  put c first (I find this a little easier to read)

~c ∧ ~(~a ∧ b)    convert to boolean algebra
~c ∧ (a ∨ ~b)     an equivalent expression

!c && (a || !b)   convert back to go code, and finally put back in substitutions

!pm.idempotencyKeyForCompleteMatch(res.CompletePromise.Promise)
  && (req.CompletePromise.Strict || pm.promise.State != promise.Timedout)
```

[Here](https://www.emathhelp.net/en/calculators/discrete-mathematics/boolean-algebra-calculator/?f=%7Ec%E2%88%A7%7E%28%7Ea%E2%88%A7b%29) is the equivalency proof, and here are the corresponding truth tables for both expressions:
- [~c ∧ ~(~a ∧ b)](https://www.emathhelp.net/en/calculators/discrete-mathematics/truth-table-calculator/?f=%7Ec+%E2%88%A7+%7E%28%7Ea+%E2%88%A7+b%29&v=)
- [~c ∧ (a ∨ ~b)](https://www.emathhelp.net/en/calculators/discrete-mathematics/truth-table-calculator/?f=%7Ec+%E2%88%A7+%28a+%E2%88%A8+%7Eb%29&v=)

Comment (favalos) [reply test/dst/model.go]: Sweet! Yes, that looks much better.

Thanks for the detailed explanation.

**Comment by favalos** [comment]
Let me know if any other comments.

**Review by dfarr — APPROVED** [review]
🎉

**Comment by dfarr** [comment]
Thanks so much again @favalos !
